sql/ir_generator.go

```diff
@@ -0,0 +1,375 @@
+// Copyright 2019 The SQLFlow Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/sql-machine-learning/sqlflow/sql/codegen"
+	"github.com/sql-machine-learning/sqlflow/sql/columns"
+)
+
+func generateTrainIR(slct *extendedSelect, connStr string) (*codegen.TrainIR, error) {
+	tc := slct.trainClause
+	estimator := tc.estimator
+	attrList, err := generateAttributeIR(&slct.trainAttrs)
+	if err != nil {
+		return nil, err
+	}
+	fcMap := make(map[string][]codegen.FeatureColumn)
+	for target, columnList := range tc.columns {
+		fcList := []codegen.FeatureColumn{}
+		for _, colExpr := range columnList {
+			if colExpr.typ != 0 {
+				// column identifier like "COLUMN a1,b1"
+				nc := &codegen.NumericColumn{
+					FieldMeta: &codegen.FieldMeta{
+						Name:      colExpr.val,
+						Shape:     []int{1},
+						DType:     codegen.Float,
+						IsSparse:  false,
+						Delimiter: "",
+					}}
+				fcList = append(fcList, nc)
+			} else {
+				fc, err := parseFeatureColumn(&colExpr.sexp)
+				if err != nil {
+					return nil, fmt.Errorf("parse FeatureColumn failed: %v", err)
+				}
+				fcList = append(fcList, fc)
+			}
+		}
+		fcMap[target] = fcList
+	}
+	label := &codegen.LabelColumn{
+		FieldMeta: &codegen.FieldMeta{
+			Name: tc.label,
+		}}
+
+	// TODO(typhoonzero): fill in ValidationSelect when VALIDATE clause is ready
```

Comment:
Where shall we put the logic of automatically training and validation splitting?

Reply:
Added TODO comment.